source/client-side-encryption/client-side-encryption.md

```diff
@@ -0,0 +1,2475 @@
+      that map onto $P$ as the `aws` property.
+4. If $K$ contains an `gcp` property, and that property is an empty map:
+   1. Attempt to obtain credentials $C$ from the environment logic as is detailed in
+      [Obtaining GCP Credentials](#obtaining-gcp-credentials).
+   2. If credentials $C$ were successfully loaded, create a new [GCPKMSOptions](#GCPKMSOptions) map from $C$ and insert
+      that map onto $P$ as the `gcp` property.
+5. If $K$ contains an `azure` property, and that property is an empty map:
+   1. If there is a `cachedAzureAccessToken` AND the duration until `azureAccessTokenExpireTime` is greater than one
+      minute, insert `cachedAzureAccessToken` as the `azure` property on $P$.
+   2. Otherwise:
+      1. Let $t_0$ be the current time.
+      2. Attempt to obtain an Azure VM Managed Identity Access Token $T$ as detailed in
+         [Obtaining an Access Token for Azure Key Vault](#obtaining-an-access-token-for-azure-key-vault).
+      3. If a token $T$ with expire duration $d\_{exp}$ were obtained successfully, create a new
+         [AzureAccessToken](#AzureAccessToken) object with $T$ as the `accessToken` property. Insert that
+         [AzureAccessToken](#AzureAccessToken) object into $P$ as the `azure` property. Record the generated
+         [AzureAccessToken](#AzureAccessToken) in `cachedAzureAccessToken`. Record the `azureAccessTokenExpireTime` as
+         $t_0 + d\_{exp}$.
+6. Return $P$ as the additional KMS providers to [libmongocrypt](#libmongocrypt).
+
+<div id="obtaining-gcp-credentials">
+
+##### Obtaining GCP Credentials
+
+Set `HOST` to `metadata.google.internal`.
+
+Send an HTTP request to the URL `http://<HOST>/computeMetadata/v1/instance/service-accounts/default/token` with the
+header `Metadata-Flavor: Google`.
+
+If the HTTP response code is not 200, return an error including the body of the HTTP response in the error message.
+
+Parse the HTTP response body as JSON. If parsing fails, return an error including the body of the HTTP response in the
+error message.
+
+Check the parsed JSON for the field "access_token". If "access_token" is not present, return an error including the body
+of the HTTP response in the error message.
+
+Return "access_token" as the credential.
+
+##### Obtaining an Access Token for Azure Key Vault
+
+Virtual machines running on the Azure platform have one or more *Managed Identities* associated with them. From within
+the VM, an identity can be used by obtaining an access token via HTTP from the *Azure Instance Metadata Service* (IMDS).
+[See this documentation for more information](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http)
+
+..note:
+
+```
+To optimize for testability, it is recommended to implement an isolated
+abstraction for communication with IMDS. This will aide in the implementation
+of the prose tests of the communication with an IMDS server.
+```
```

Comment:
This should be changed to a note blockquote.

Reply:
Done